---
title: "Aligning_with_public_database"
output: html_document
---

Libraries
```{r}
library(Biostrings)
library(scifer)
library(DECIPHER)
```

```{r}
#convert pladbd into a .fasta file
plabdab <- read.csv("~/Documents/Sequencing/HW_malaria_5002_3001_2008_1010/public_database/plabdab.csv", sep = ",")
my_seqs_df <- read.csv("~/Documents/Sequencing/HW_malaria_5002_3001_2008_1010/clonotypes/final/merged_file_with_5013.tsv", sep = "\t")

# Assuming the heavy chain sequences are in column 'Heavy_VDJ'
# Create a temporary dataframe for exporting
df <- data.frame(
  id = paste0("plabdab_seq_", seq_along(plabdab$heavy_sequence)),
  sequence = plabdab$heavy_sequence,
  stringsAsFactors = FALSE
)

df_to_fasta(
  sequence_name = plabdab$ID,
  sequence_strings = plabdab$heavy_sequence,
  file_name = "plabdab_sequences.fasta",
  output_dir = "~/Documents/Sequencing/HW_malaria_5002_3001_2008_1010/public_database/",
  save_fasta = TRUE
)

#df to sequences for my_seq
df_to_fasta(
  sequence_name = my_seqs_df$sequence_id,
  sequence_strings = my_seqs_df$sequence_alignment_aa,
  file_name = "my_aa_sequences.fasta",
  output_dir = "~/Documents/Sequencing/HW_malaria_5002_3001_2008_1010/output_data_general/",
  save_fasta = TRUE
)
```
Trying to align the sequences from public database to my sequences
```{r}
# Read sequences from a FASTA file (your own and PLAbDab)
plabdab_seqs <- readDNAStringSet("~/Documents/Sequencing/HW_malaria_5002_3001_2008_1010/public_database/plabdab_sequences.fasta")
my_seqs <- readDNAStringSet("~/Documents/Sequencing/HW_malaria_5002_3001_2008_1010/output_data_general/my_aa_sequences.fasta")

# Print the sequences to verify
plabdab_seqs
my_seqs
```

```{r}
# Align your sequences to PLAbDab sequences (pairwise)
alignment_results <- pairwiseAlignment(my_seqs[1], plabdab_seqs[1], substitutionMatrix = "BLOSUM62")

# Print alignment results
alignment_results


alignment_results <- pairwiseAlignment(my_seqs["3001_01_90-1177654324_ab1_B12"], plabdab_seqs["URF29151_URF29152"], substitutionMatrix = "BLOSUM62")
```
Multiple sequence alignment
```{r}
# Combine both sets of sequences into a single dataset
all_seqs <- c(plabdab_seqs, my_seqs)

# Perform multiple sequence alignment
aligned_seqs <- AlignSeqs(all_seqs)

# View the alignment
BrowseSeqs(aligned_seqs)

```

best match
```{r}
# Create a dataframe to store results
results <- data.frame(
  Your_Seq_ID = character(),
  Best_Match_ID = character(),
  Identity_Percent = numeric(),
  stringsAsFactors = FALSE
)

# For each of your sequences, find the most similar one in PLAbDab
for (i in seq_along(my_seqs)) {
  my_seq <- my_seqs[[i]]
  best_identity <- 0
  best_id <- ""

  for (j in seq_along(plabdab_seqs)) {
    alignment <- pairwiseAlignment(
      my_seq,
      plabdab_seqs[[j]],
      substitutionMatrix = "BLOSUM62",
      gapOpening = 10,
      gapExtension = 0.5
    )

    identity <- as.numeric(pid(alignment)[1])

    if (identity > best_identity) {
      best_identity <- identity
      best_id <- names(plabdab_seqs)[j]
    }
  }

  results <- rbind(results, data.frame(
    Your_Seq_ID = names(my_seqs)[i],
    Best_Match_ID = best_id,
    Identity_Percent = best_identity
  ))
}

#save results
write.csv(results, "~/Documents/Sequencing/HW_malaria_5002_3001_2008_1010/output_data_general/identity_results.csv", row.names = FALSE)
```


summarise the most common abs that match my sequences
```{r}
df_results_summary <- results %>%
  group_by(Best_Match_ID) %>%
  summarise(average_identity = mean(Identity_Percent, na.rm = TRUE))
```

