---
title: "analysis of qc for smartseq files"
output: html_document
---
Libraries
```{r}
library(ggplot2)
library(dplyr)
library(ggpubr)
library(stringr)
```
Data upload
```{r}
airr_data <- read.csv("~/Documents/Sequencing/Smart_seq/BCR_analysis/smartseq_airr.tsv", sep = "\t")
igd_igblast_smartseq_data <- read.csv("~/Documents/Sequencing/Smart_seq/BCR_analysis/Results/hc_smart_seq_IgD/final/assigned.tsv", sep = "\t")
hc_sanger_data <- read.csv("~/Documents/Sequencing/250317_hc_igdiscover/final/assigned.tsv", sep = "\t")
lc_sanger_data <- read.csv("~/Documents/Sequencing/250317_lc_igdiscover/final/assigned.tsv", sep = "\t")
all_sanger_data <- rbind(hc_sanger_data,lc_sanger_data)

plate_map <- read.csv("~/Documents/Sequencing/Smart_seq/plate_map.csv", sep = ";")

#data cleaning and annotation 
igd_igblast_smartseq_data <- igd_igblast_smartseq_data %>%
  mutate(Well = str_extract(sequence_id, "^[^_]+")) %>%
  left_join(plate_map, by = "Well")

saveRDS(igd_igblast_smartseq_data, file = "~/Documents/Sequencing/Smart_seq/igd_igblast_smartseq_data.rds")

airr_data<- airr_data %>%
  mutate(Well = str_extract(sequence_id, "^[^_]+")) %>%
  left_join(plate_map, by = "Well")

saveRDS(airr_data, file = "~/Documents/Sequencing/Smart_seq/airr_data.rds")

airr_data_family_allocation_IGH <- airr_data %>%
  dplyr::select(Well, v_call, d_call, j_call, locus) %>%
  mutate(v_family_H = substr(v_call, 1, 5),
         d_family_H = substr(d_call, 1, 5),
         j_family_H = substr(j_call, 1, 5)) %>%
  filter(locus == "IGH")

saveRDS(airr_data_family_allocation_IGH, file = "~/Documents/Sequencing/Smart_seq/airr_data_family_allocation_IGH.rds")

airr_data_family_allocation_light <- airr_data %>%
  dplyr::select(Well, v_call, d_call, j_call, locus) %>%
  mutate(v_family_L = substr(v_call, 1, 5),
         j_family_L = substr(j_call, 1, 5)) %>%
  filter(locus != "IGH")

saveRDS(airr_data_family_allocation_light, file = "~/Documents/Sequencing/Smart_seq/airr_data_family_allocation_light.rds")

family_allocation <- airr_data_family_allocation_IGH %>%
  full_join(airr_data_family_allocation_light, by = "Well") %>%
  dplyr::select(Well, v_family_H, d_family_H, j_family_H, v_family_L, j_family_L)


saveRDS(family_allocation, file = "~/Documents/Sequencing/Smart_seq/family_allocation.rds")

plate_map <- plate_map %>%
  full_join(family_allocation, by = "Well")
```

Plotting somatic hypermutation sanger vs. smartseq
```{r}
sh_sanger <- all_sanger_data %>%
  select(sequence_id,locus,V_SHM,J_SHM)

sh_sanger$method <- "sanger"

sh_smartseq <- igd_igblast_smartseq_data %>%
  select(sequence_id,locus,V_SHM,J_SHM)

sh_smartseq$method <- "smartseq"

sh_all <- rbind(sh_sanger, sh_smartseq)
comparisons <- list(c("sanger", "smartseq"))

sh_all_plot <- ggplot(sh_all, aes(x = method, y = V_SHM, fill = method)) +
  geom_violin(trim = FALSE, position = position_dodge(width = 0.8)) +
  geom_boxplot(width = 0.1, position = position_dodge(width = 0.8), outlier.shape = NA)  +
  stat_compare_means(
    method = "wilcox.test",
    comparisons = comparisons,
    label = "p.format",
    hide.ns = TRUE
  ) +
  facet_wrap(~locus) +
  theme_minimal() +
  labs(x = "", y = "V_SHM") +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 45, hjust = 1)
  )
sh_all_plot
ggsave("~/Documents/Thesis_data/sh_all_plot.png", plot = sh_all_plot, width = 8, height = 6, dpi = 300)


```

compare v_identity
```{r}
v_identity_sanger <- all_sanger_data %>%
  select(sequence_id,locus,v_identity)
v_identity_sanger$method <- "sanger"
v_identity_smartseq <- igd_igblast_smartseq_data %>%
  select(sequence_id,locus,v_identity)
v_identity_smartseq$method <- "smartseq"
v_identity_all <- rbind(v_identity_sanger, v_identity_smartseq)

comparisons <- list(c("sanger", "smartseq"))

v_identity_all_plot <- ggplot(v_identity_all, aes(x = method, y = v_identity, fill = method)) +
  geom_violin(trim = FALSE, position = position_dodge(width = 0.8)) +
  geom_boxplot(width = 0.1, position = position_dodge(width = 0.8), outlier.shape = NA)  +
  stat_compare_means(
    method = "wilcox.test",
    comparisons = comparisons,
    label = "p.format",
    hide.ns = TRUE
  ) +
  facet_wrap(~locus) +
  theme_minimal() +
  labs(x = "", y = "V Identity") +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

v_identity_all_plot
ggsave("~/Documents/Thesis_data/v_identity_all_plot.png", plot = v_identity_all_plot, width = 8, height = 6, dpi = 300)
```
j_identity score
```{r}
j_identity_sanger <- all_sanger_data %>%
  select(sequence_id,locus,j_identity)
j_identity_sanger$method <- "sanger"
j_identity_smartseq <- igd_igblast_smartseq_data %>%
  select(sequence_id,locus,j_identity)
j_identity_smartseq$method <- "smartseq"
j_identity_all <- rbind(j_identity_sanger, j_identity_smartseq)
comparisons <- list(c("sanger", "smartseq"))
j_identity_all_plot <- ggplot(j_identity_all, aes(x = method, y = 
                                                    j_identity, fill = method)) +
  geom_violin(trim = FALSE, position = position_dodge(width = 0.8)) +
  geom_boxplot(width = 0.1, position = position_dodge(width = 0.8), outlier.shape = NA)  +
  stat_compare_means(
    method = "wilcox.test",
    comparisons = comparisons,
    label = "p.format",
    hide.ns = TRUE
  ) +
  facet_wrap(~locus) +
  theme_minimal() +
  labs(x = "", y = "J Identity") +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 45, hjust = 1)
  )
j_identity_all_plot
ggsave("~/Documents/Thesis_data/j_identity_all_plot.png", plot = j_identity_all_plot, width = 8, height = 6, dpi = 300)

```
Counting occurance of light chain heavy chain pairs
```{r}
#smartseq
seq_id_counts <- table(airr_data$cell_id)
# Filter IDs that appear more than once and count them
num_duplicates <- sum(seq_id_counts > 1 & seq_id_counts < 3)

# Print the result
print(num_duplicates)

all_sequences <- airr_data %>%
  filter(locus == "IGH") %>%
  group_by(Well) %>%
  summarise(count = n())

total_smartseq <- sum(all_sequences$count)

percentage_occurance_smartseq <- (num_duplicates / total_smartseq) * 100
print(percentage_occurance_smartseq)

#sanger
dictionary_sanger <- read.csv("~/Documents/Sequencing/heavy_chain_light_chain_dictionary.csv", sep = ";")
counting_matches_data <- all_sanger_data %>%
  dplyr::select(sequence_id)
#counting_matches_data$heavy_chain_name <- ""

# Join on matching sequence_id and light_chain
counting_matches_data <- counting_matches_data %>%
  left_join(dictionary_sanger, by = c("sequence_id" = "Light_chain"))

num_duplicates_sanger <- length(unique(na.omit(counting_matches_data$Heavy_chain)))

seq_id_counts_sanger <- table(counting_matches_data$Heavy_chain)
# Filter IDs that appear more than once and count them
more_than_one_light_chain <- sum(seq_id_counts_sanger > 1 & seq_id_counts_sanger < 3)

counting_matches_data <- counting_matches_data %>%
  filter(is.na(Heavy_chain)) %>%
  group_by(sequence_id) %>%
  summarise(count = n())
total_sanger <- sum(counting_matches_data$count)
# Print the result
print(more_than_one_light_chain)

percentage_occurance_sanger <- (num_duplicates_sanger / total_sanger) * 100
print(percentage_occurance_sanger)


```

See distribution just based on blood_CSP
```{r}
airr_data_bCSP_only <- airr_data %>%
  filter(Sample_type == "Blood_CSP")
#smartseq
seq_id_counts_bCSP <- table(airr_data_bCSP_only$cell_id)
# Filter IDs that appear more than once and count them
num_duplicates_bCSP <- sum(seq_id_counts_bCSP > 1 & seq_id_counts_bCSP < 3)
# Print the result
print(num_duplicates_bCSP)

all_sequences_blood_CSP_only <- airr_data_bCSP_only %>%
  filter(locus == "IGH") %>%
  group_by(Well) %>%
  summarise(count = n())

total_smartseq_bCSP <- sum(all_sequences_blood_CSP_only$count)
print(total_smartseq_bCSP)

percentage_occurance_smartseq <- (num_duplicates_bCSP / total_smartseq_bCSP) * 100
print(percentage_occurance_smartseq)
```

```{r}
colnames(airr_data)
```
```{r}
colnames(igd_igblast_smartseq_data)
```
V_covered - Percentage alignment score
```{r}
v_covered_sanger <- all_sanger_data %>%
  select(sequence_id,locus,V_covered)
v_covered_sanger$method <- "sanger"
v_covered_smartseq <- igd_igblast_smartseq_data %>%
  select(sequence_id,locus,V_covered)
v_covered_smartseq$method <- "smartseq"
v_covered_all <- rbind(v_covered_sanger, v_covered_smartseq)
comparisons <- list(c("sanger", "smartseq"))

v_covered_all_plot <- ggplot(v_covered_all, aes(x = method, y = V_covered, fill = method)) +
  # geom_violin(trim = FALSE, position = position_dodge(width = 0.8)) +
  geom_boxplot(width = 0.5, position = position_dodge(width = 0.8), outlier.shape = NA)  +
  stat_compare_means(
    method = "wilcox.test",
    comparisons = comparisons,
    label = "p.format",
    hide.ns = TRUE
  ) +
  scale_fill_lancet() +
  facet_wrap(~locus) +
  theme_bw() +
  labs(x = "", y = "Alignment coverage (%)") +
  theme(aspect.ratio = 1,
    legend.position = "none",
    #axis.text.x = element_text(angle = 45, hjust = 1, size = 14),
    #axis.text.y = element_text(size = 14),
    #axis.title.y = element_text(size = 16),
    #strip.text = element_text(size = 16, face = "bold"),
    #plot.title = element_text(size = 18, face = "bold", hjust = 0.5)  # fixed typo here
  ) +
  coord_cartesian(ylim = c(85, 110))

v_covered_all_plot
ggsave("~/Documents/Thesis_data/v_covered_all_plot.png", plot = v_covered_all_plot, width = 8, height = 6, dpi = 300)
```
D_covered
```{r}
d_covered_sanger <- all_sanger_data %>%
  select(sequence_id,locus,D_covered) %>%
  filter(locus == "IGH")
d_covered_sanger$method <- "sanger"
d_covered_smartseq <- igd_igblast_smartseq_data %>%
  select(sequence_id,locus,D_covered) %>%
  filter(locus == "IGH")
d_covered_smartseq$method <- "smartseq"
d_covered_all <- rbind(d_covered_sanger, d_covered_smartseq)
comparisons <- list(c("sanger", "smartseq"))

d_covered_all_plot <- ggplot(d_covered_all, aes(x = method, y = D_covered, fill = method)) +
  #geom_violin(trim = FALSE, position = position_dodge(width = 0.8)) +
  geom_boxplot(width = 0.5, position = position_dodge(width = 0.8), outlier.shape = NA)  +
  stat_compare_means(
    method = "wilcox.test",
    comparisons = comparisons,
    label = "p.format",
    hide.ns = TRUE
  ) +
  facet_wrap(~locus) +
  theme_minimal() +
  labs(x = "", y = "Alignment coverage (%)", title = "D-gene") +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 45, hjust = 1, size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.y = element_text(size = 16),
    strip.text = element_text(size = 16, face = "bold"),
    plot.title = element_text(size = 18, face = "bold", hjust = 0.5)  # fixed typo here
  ) +#+
  #coord_cartesian(ylim = c(75, 110)) 
  #scale_y_break(c(5, 90)) +
  coord_cartesian(ylim = c(0, 110))
d_covered_all_plot
ggsave("~/Documents/Thesis_data/d_covered_all_plot.png", plot = d_covered_all_plot, width = 8, height = 6, dpi = 300)
```
J_covered
```{r}
j_covered_sanger <- all_sanger_data %>%
  select(sequence_id,locus,J_covered)
j_covered_sanger$method <- "sanger"
j_covered_smartseq <- igd_igblast_smartseq_data %>%
  select(sequence_id,locus,J_covered)
j_covered_smartseq$method <- "smartseq"
j_covered_all <- rbind(j_covered_sanger, j_covered_smartseq)
comparisons <- list(c("sanger", "smartseq"))
j_covered_all_plot <- ggplot(j_covered_all, aes(x = method, y = J_covered, fill = method)) +
  #geom_violin(trim = FALSE, position = position_dodge(width = 0.8)) +
  geom_boxplot(width = 0.5, position = position_dodge(width = 0.8), outlier.shape = NA)  +
  stat_compare_means(
    method = "wilcox.test",
    comparisons = comparisons,
    label = "p.format",
    hide.ns = TRUE
  ) +
  facet_wrap(~locus) +
  theme_minimal() +
  labs(x = "", y = "Alignment coverage (%)", title = "J gene") +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 45, hjust = 1, size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.y = element_text(size = 16),
    strip.text = element_text(size = 16, face = "bold"),
    plot.title = element_text(size = 18, face = "bold", hjust = 0.5)  # fixed typo here
  ) +
  #coord_cartesian(ylim = c(75, 110))
  #scale_y_break(c(5, 90)) +
  coord_cartesian(ylim = c(70, 110))
j_covered_all_plot
ggsave("~/Documents/Thesis_data/j_covered_all_plot.png", plot = j_covered_all_plot, width = 8, height = 6, dpi = 300)
```
Putting all plots onto the same plot
```{r}
library(patchwork)

library(patchwork)

percentage_alignment_all_plot <- (v_covered_all_plot | d_covered_all_plot | j_covered_all_plot) +
  plot_layout(guides = "collect") +   # << this line is important
  plot_annotation(
    title = "Percentage Alignment for V, D, and J Genes",
    tag_levels = "A"
  ) #&
  #theme_bw()



percentage_alignment_all_plot

ggsave("~/Documents/Thesis_data/percentage_alignment_all_plot.png", plot = percentage_alignment_all_plot, width = 8, height = 6, dpi = 300)
```


Plotting V_quality vs alignment coverage for igblast smartseq data
```{r}
airr_data_plot_V_SHM <- ggplot(igd_igblast_smartseq_data, aes(x = V_SHM, y = V_covered, color = locus, shape = Sample_type)) +
  geom_point(alpha = 0.7) +
  scale_x_log10() +  # count often spans orders of magnitude
  theme_minimal() +
  labs(title = "Alignment coverage(%) vs. SHM for IGDiscover smartseq data",
       x = "V_SHM",
       y = "Alignment coverage (%)",
       color = "Chain Type")

airr_data_plot_V_SHM
```
```{r}
# Pearson correlation
cor.test(igd_igblast_smartseq_data$V_SHM, igd_igblast_smartseq_data$V_covered, method = "pearson")
cor.test(igd_igblast_smartseq_data$V_SHM, igd_igblast_smartseq_data$V_covered, method = "spearman")
```

